### PR TITLE
AT-4007 Add Flappy Lambda

### PR DESCRIPTION
**Pull request by jconstance-amplify**
First pass on the code.

ToDo:
- [x] Unit Tests

### CODE REVIEW

File: flappy_detector/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+"""Models used by Flappy Detector"""
+from .flappy_event import FlappyEvent
```
Comment (aoskotsky-amplify) [inline flappy_detector/models/__init__.py:2]: Is this supposed to be a relative import?

File: flappy_detector/handlers/detect.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        self._send_alerts(flapping_events=flapping_events)
+
+    def _get_events(self) -> List[Dict[str, Any]]:
+        """
+        Get all relevant DynamoDB records.
+        :return: List of all DynamoDB records, no older than the max_event_age.
+        """
+        cut_off = Decimal((datetime.now() - self.max_event_age).timestamp())
+        return get_boto3_paged_results(
+            self.dynamodb_table.scan,
```
Comment (aoskotsky-amplify) [inline flappy_detector/handlers/detect.py:88]: Could this have been a query?

Comment (jconstance-amplify) [reply flappy_detector/handlers/detect.py:88]: No.

**Review by aoskotsky-amplify — APPROVED** [review]

File: flappy_detector/handlers/detect.py (hunk 2 above)
Comment (aoskotsky-amplify) [reply flappy_detector/handlers/detect.py:88]: Very well then

File: flappy_detector/handlers/ingest.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        :return: The CloudWatch events grouped first by account, then by region.
+        """
+        grouped_events: Dict[str, Dict[str, List[Dict[str, Any]]]] = defaultdict(lambda: defaultdict(list))
+        for event in events:
+            grouped_events[event["account"]][event["region"]].append(
+                {
+                    "instance_id": event["detail"]["instance-id"],
+                    "state": event["detail"]["state"],
+                    "timestamp": Decimal(parse(event["time"]).timestamp()),
+                    "region": event["region"],
```
Comment (jconstance-amplify) [inline flappy_detector/handlers/ingest.py:82]: Can probably remove this without too much effort.

File: flappy_detector/handlers/detect.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                flappy_event = FlappyEvent(
+                    account=event["account"],
+                    region=event["region"],
+                    environment=event["environment"],
+                    application=event["application"],
+                    group_name=event["group_name"],
+                )
+                flappy_event = flappy_events.get(flappy_event.key, flappy_event)
+            except TypeError:
+                logger.warning(
```
Comment (jconstance-amplify) [inline flappy_detector/handlers/detect.py:115]: Could make it an error

File: flappy_detector/handlers/detect.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                        "event": event,
+                    }
+                )
+                continue
+
+            if not flappy_event.team:
+                flappy_event.team = event.get("team")
+
+            flappy_event.count += 1
+            flappy_event.spread += Ec2State(event["state"]).change
```
Comment (jconstance-amplify) [inline flappy_detector/handlers/detect.py:127]: Try catching this to prevent one bad dynamodb entry from breaking everything.

**Review by rddimon — APPROVED** [review]

**Pull request merged**